libsolidity/ast/Types.cpp

```diff
@@ -3472,15 +3470,19 @@ Type const* FunctionType::mobileType() const
 	if (valueSet() || gasSet() || saltSet() || hasBoundFirstArgument())
 		return nullptr;
 
-	// return function without parameter names
+	// Special function types do not get a mobile type, such that they cannot be used in complex expressions.
+	if (m_kind != FunctionType::Kind::Internal && m_kind != FunctionType::Kind::External && m_kind != FunctionType::Kind::DelegateCall)
+		return nullptr;
+
+	// return function without parameter names and without declaration
 	return TypeProvider::function(
 		m_parameterTypes,
 		m_returnParameterTypes,
 		strings(m_parameterTypes.size()),
 		strings(m_returnParameterNames.size()),
 		m_kind,
 		m_stateMutability,
-		m_declaration,
+		nullptr,
```

Comment:
The PR now fails on a case added in #14365. Expectations need to be updated.

Though if we do that, the test case will no longer be testing the newly added error. I wonder if we can even still trigger it. Perhaps we should replace it with an assert?

I'd check if `D.f == D.g` will trigger this implicit conversion.

Reply:
```
TypeError 2271: (131-141): Built-in binary operator == cannot be applied to types function D.f() and function D.g().
```

I've already updated the expectations, but it's a good point you make. Not sure how to trigger the original error now?

Reply:
Let's first try to replace it with an assert. If it fails, the answer will be obvious. If it doesn't, maybe it's really impossible.